Implement rfc6587 auto detection

bazsi opened the pull request:
This is a pretty common misconfiguration that should be handled automatically. 

Just use syslog(transport(auto)) to automatically accept both octet counted and non-octet counted connections, this opens a TCP port without TLS, meaning it is an implementation of RFC6587.

alltilla commented on lib/logproto/logproto-auto-server.c:
```diff
@@ -0,0 +1,147 @@
+  else if (rc < 0)
+    return LPS_ERROR;
+
+  self->proto_impl = _construct_detected_proto(self, detect_buffer, rc);
+  if (self->proto_impl)
+    {
+      /* transport is handed over to the new proto */
+      self->super.transport = NULL;
+      return LPS_SUCCESS;
+    }
```
Shouldn't we do a handshake after we autodetected the proto?

alltilla commented on news/feature-4814.md:
```diff
@@ -0,0 +1,5 @@
+octet-count based framing to avoid confusion that stems from the sender
+using a different protocol to the server.  This is automatically enabled for
+both the `transport(tcp)` and the `transport(tls)` settings.  You can disable
+auto-detection by using `transport(text)` or `transport(framed)`.
```
We disable auto detection with `transport(force-framed)`, right?

bazsi commented:
This should wait until #156 is merged.

bazsi commented:
I have now rebased it but still need to check if it works on top of #156

alltilla commented:
#156 is now merged, I think we can give this another go (rebase, CI, review, ...)

bazsi commented:
This should now be ready for review.

Note: there's a followup PR as well, #361

bazsi commented:
I have now opted out of doing this by default. The auto-detection mechanism is only enabled with transport(auto)

MrAnno commented:
I've started reviewing this patchset, will finish it by the end of the week.

OverOrion reviewed (commented):
I've reviewed this and tested it manually as well, worked just fine :rocket:.
Is there a reason why it's still marked as `WIP`?

bazsi commented:
I pulled some of the later commits into this one, rebased against main and removed WIP.

@OverOrion if you can please re-approve, but note that I did pull in some newer changes into this.
